providers/ms_entra_id.go

```diff
@@ -125,29 +127,79 @@ func (p *MicrosoftEntraIDProvider) redeemWithFederatedToken(ctx context.Context,
 	params.Add("code", code)
 	params.Add("grant_type", "authorization_code")
 
-	// perform exchange
-	resp := requests.New(p.RedeemURL.String()).
-		WithContext(ctx).
-		WithMethod("POST").
-		WithBody(bytes.NewBufferString(params.Encode())).
-		SetHeader("Content-Type", "application/x-www-form-urlencoded").
-		Do()
+	token, err := p.fetchToken(ctx, params)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching token: %w", err)
+	}
 
-	// prepare token of type *oauth2.Token
-	var token *oauth2.Token
-	var rawResponse interface{}
+	return p.OIDCProvider.createSession(ctx, token, false)
+}
 
-	body := resp.Body()
-	if err := json.Unmarshal(body, &rawResponse); err != nil {
-		return nil, err
+// RefreshSession uses the RefreshToken to fetch new Access and ID Tokens
+func (p *MicrosoftEntraIDProvider) RefreshSession(ctx context.Context, s *sessions.SessionState) (bool, error) {
+	if s == nil || s.RefreshToken == "" {
+		return false, nil
 	}
 
-	if err := json.Unmarshal(body, &token); err != nil {
-		return nil, err
+	var err error
+	ctx = oidc.ClientContext(ctx, requests.DefaultHTTPClient)
+	if p.federatedTokenAuth {
+		err = p.redeemRefreshTokenWithFederatedToken(ctx, s)
+	} else {
+		err = p.redeemRefreshToken(ctx, s)
 	}
 
-	// create session using new token and generic OIDC provider
-	return p.OIDCProvider.createSession(ctx, token.WithExtra(rawResponse), false)
+	if err != nil {
+		return false, fmt.Errorf("unable to redeem refresh token: %v", err)
+	}
+
+	return true, nil
+}
+
+// redeemRefreshTokenWithFederatedToken uses a RefreshToken and federacted credentials with the RedeemURL to refresh the
+// Access Token and (probably) the ID Token.
```

Comment:
Does it refresh the ID token or not?

Reply:
What does the Entra ID documentation say about this?

Reply:
The example does show it returns a ID Token, but since the spec sais its optional I don't want to trust it to much (it will probably never change though)

https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-auth-code-flow#refresh-the-access-token